src/WCSimIBDGen.cc

```diff
@@ -0,0 +1,258 @@
+
+WCSimIBDGen::~WCSimIBDGen() {
+    // Delete things here
+}
+
+void WCSimIBDGen::ReadSpectrumFromDB(G4String spectrum_name, std::string model_name) {
+    // Open the database
+    std::ifstream spectrum_json(spectrum_name);
+
+    // Check that the database has been found
+    if (!spectrum_json.is_open()) {
+        G4cerr << "IBDGen: [ERROR] spectrum database " << spectrum_name << " not found." << G4endl;
+        exit(-1);
+    } else {
+        G4cout << "IBDGen: [INFO] spectrum database " << spectrum_name << " found." << G4endl;
+    }
+
+    std::stringstream buffer;
+    buffer << spectrum_json.rdbuf();
+
+    json data = json::parse(buffer.str());
+
+    // Loop over the json
+    for (const auto &model : data["models"]) {
+        if (model["name"].get<string>() == model_name) {
+            G4cout << "IBDGen: [INFO] using model " << model_name << G4endl;
+
+            energy = model["energy"].get<std::vector<float>>();
+            flux = model["flux"].get<std::vector<float>>();
+
+            G4cout << "IBDGen: [INFO] spectrum read from database" << G4endl;
+
+            // Check if the flux vector and energy vector have the same size
+            if (energy.size() != flux.size()) {
+                G4cerr << "IBDGen: \033[31m[ERROR]\033[0m energy and flux vectors have different sizes. Energy: "
+                       << energy.size() << ", flux: " << flux.size() << ". Check " << spectrum_name << G4endl;
+                exit(-1);
+            }
+            // Set the minimum and maximum energy
+            e_min = energy.front();
+            e_max = energy.back();
+
+            // Set the maximum flux
+            flux_max = *std::max_element(flux.begin(), flux.end());
+
+            return;
+        }
+    }
+
+    // If we get here, the model was not found
+    G4cerr << "IBDGen: \033[31m[ERROR]\033[0m model \033[31m" << model_name << "\033[0m not found in database "
+           << spectrum_name << G4endl;
+    exit(-1);
+    return;
+}
+
+double WCSimIBDGen::InterpolateSpectrum(std::vector<float> ener_vec, std::vector<float> flux_vec, float ene) {
+
+    // Interpolate the spectrum at the energies in ener_vec
+    // The spectrum is given by the vectors ener_vec and flux_vec
+
+    // Loop over the energies in energy
+    for (size_t i = 1; i < ener_vec.size(); i++) {
+        if (ene <= ener_vec[i]) {
+            // Perform linear interpolation
+            double e1 = ener_vec[i - 1];
+            double e2 = ener_vec[i];
+            double f1 = flux_vec[i - 1];
+            double f2 = flux_vec[i];
+
+            double interpolated_value = f1 + (f2 - f1) * (ene - e1) / (e2 - e1);
+            return interpolated_value;
+        }
+    }
+
+    // If energy is larger than the maximum energy in the spectrum, return the final flux value
+    return flux_vec.back();
+}
+
+G4ThreeVector WCSimIBDGen::GenRandomPosition() {
+    // Generate random neutrino position
+    // Pick random position in detector
```

Comment:
Can you be clear about what "detector" means here? Is it the cuboidal world volume? The cylindrical extent of the tank? Just inside the ID? Something else?

Reply:
Added a comment about this referring to the inner detector.